tests/browser/test_tabhistory.py

```diff
@@ -0,0 +1,139 @@
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for webelement.tabhistory."""
+
+from PyQt5.QtCore import QUrl, QPoint
+import pytest
+
+from qutebrowser.browser import tabhistory
+from qutebrowser.browser.tabhistory import TabHistoryItem as Item
+from qutebrowser.utils import qtutils
+
+
+class TestSerializeHistory:
+
+    """Tests for serialize()."""
+
+    ITEMS = [
+        Item(QUrl('https://www.heise.de/'), QUrl('http://www.heise.de/'),
+             'heise'),
+        Item(QUrl('http://example.com/%E2%80%A6'),
+             QUrl('http://example.com/%E2%80%A6'), 'percent', active=True),
+        Item(QUrl('http://example.com/?foo=bar'),
+             QUrl('http://original.url.example.com/'), 'arg',
+             user_data={'foo': 23, 'bar': 42}),
+        # From https://github.com/OtterBrowser/otter-browser/issues/709#issuecomment-74749471
+        Item(
+            QUrl('http://github.com/OtterBrowser/24/134/2344/otter-browser/'
+                 'issues/709/'),
+            QUrl('http://github.com/OtterBrowser/24/134/2344/otter-browser/'
+                 'issues/709/'),
+            'Page not found | github',
+            user_data={'zoom': 149, 'scroll-pos': QPoint(0, 0)}),
+        Item(
+            QUrl('https://mail.google.com/mail/u/0/#label/some+label/'
+                 '234lkjsd0932lkjf884jqwerdf4'),
+            QUrl('https://mail.google.com/mail/u/0/#label/some+label/'
+                 '234lkjsd0932lkjf884jqwerdf4'),
+            '"some label" - email@gmail.com - Gmail"',
+            user_data={'zoom': 120, 'scroll-pos': QPoint(0, 0)}),
+    ]
+
+    @pytest.fixture(autouse=True)
+    def setup(self, webpage):
+        self.page = webpage
+        self.history = self.page.history()
+        assert self.history.count() == 0
+
+        stream, _data, self.user_data = tabhistory.serialize(self.ITEMS)
+        qtutils.deserialize_stream(stream, self.history)
+
+    def test_count(self):
+        """Check if the history's count was loaded correctly."""
+        assert self.history.count() == len(self.ITEMS)
+
+    @pytest.mark.parametrize('i', range(len(ITEMS)))
+    def test_valid(self, i):
+        """Check if all items are valid."""
+        assert self.history.itemAt(i).isValid()
+
+    @pytest.mark.parametrize('i', range(len(ITEMS)))
+    def test_no_userdata(self, i):
+        """Check if all items have no user data."""
+        assert self.history.itemAt(i).userData() is None
+
+    def test_userdata(self):
+        """Check if all user data has been restored to self.user_data."""
+        for item, user_data in zip(self.ITEMS, self.user_data):
```

Comment:
Oh, totally agree, didn't realize that at the time of the conversion. :smile:

Had to do it slightly different though:

```python
    def test_userdata(self):
        """Check if all user data has been restored to self.user_data."""
        userdata_items = [item.user_data for item in self.ITEMS]
        assert userdata_items == self.user_data
```